### PR TITLE
XD-3680: Consistent Stream.undeployed state

### PR DESCRIPTION
**Pull request by ericbottard**
fixes spring-cloud/spring-cloud-dataflow#152 

This is a first step of things discussed with @markfisher and @pperalta.
The task `complete` enum value is not taken care of in this PR. Primarily looking for feedback on: 
* javadoc and API contract expressed in the new top level enum. 
* logic implemented in `reduce()`

### CODE REVIEW

File: spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/DeploymentState.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Deployment states for modules and streams. Unless indicated, these states
+ * may represent the state of<ul>
+ * <li>an entire stream</li>
+ * <li>the global state of a deployed module as part of a stream</li>
+ * <li>the state of a particular instance of a module, in cases where {@literal module.count > 1}</li>
```
Comment (pperalta) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/DeploymentState.java]: For this JavaDoc, was the intent for module.count to be in fixed width font or the same font as the rest of the text?

Comment (ericbottard) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/DeploymentState.java]: Fixed. Did i screw up?

Comment (pperalta) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/DeploymentState.java]: Here's the difference between `{@literal}` and `{@code}`:

https://blogs.oracle.com/darcy/entry/javadoc_tip_code_and_literal

I'll modify this on merge.

File: spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/ModuleDeployer.java [hunk 2]
```diff
@@ -69,7 +69,7 @@
 	 *
 	 * @param id id for the module this status is for
 	 *
-	 * @return module deployment status
+	 * @return module deployment status or {@literal null} if this deployer knows nothing about the given deployment
```
Comment (pperalta) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/ModuleDeployer.java]: Is this accurate? I think we're supposed to return `unknown` for unknown deployments.

Comment (mbogoevici) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/ModuleDeployer.java]: I don't think we should return `null` either, also see https://github.com/spring-cloud/spring-cloud-dataflow/pull/172#discussion_r46330735

Comment (ericbottard) [inline spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/ModuleDeployer.java]: yes. This was an earlier approach. Changed my mind when we started discussing.

**Comment by pperalta** [comment]
Due to rebasing, will close this PR and replace with https://github.com/spring-cloud/spring-cloud-dataflow/pull/205.

**Pull request closed without merging**